httpclient5/src/main/java/org/apache/hc/client5/http/impl/async/AsyncHttpRequestRetryExec.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.hc.core5.http.HttpException;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.hc.core5.http.nio.AsyncDataConsumer;
+import org.apache.hc.core5.http.nio.AsyncEntityProducer;
+import org.apache.hc.core5.util.Args;
+import org.apache.hc.core5.util.TimeValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Request executor in the asynchronou srequest execution chain that is
+ * responsible for making a decision whether a request than failed due to
+ * an I/O error or received specific response from the target server should
+ * be re-executed.
+ * <p>
+ * Further responsibilities such as communication with the opposite
+ * endpoint is delegated to the next executor in the request execution
+ * chain.
+ * </p>
+ *
+ * @since 5.0
+ */
+@Contract(threading = ThreadingBehavior.STATELESS)
+@Internal
+public final class AsyncHttpRequestRetryExec implements AsyncExecChainHandler {
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    private final HttpRequestRetryStrategy retryStrategy;
+
+    public AsyncHttpRequestRetryExec(final HttpRequestRetryStrategy retryStrategy) {
+        Args.notNull(retryStrategy, "retryStrategy");
+        this.retryStrategy = retryStrategy;
+    }
+
+    private void internalExecute(
+            final int execCount,
+            final HttpRequest request,
+            final AsyncEntityProducer entityProducer,
+            final AsyncExecChain.Scope scope,
+            final AsyncExecChain chain,
+            final AsyncExecCallback asyncExecCallback) throws HttpException, IOException {
+
+        final String exchangeId = scope.exchangeId;
+
+        chain.proceed(RequestCopier.INSTANCE.copy(request), entityProducer, scope, new AsyncExecCallback() {
+
+            @Override
+            public AsyncDataConsumer handleResponse(
+                    final HttpResponse response,
+                    final EntityDetails entityDetails) throws HttpException, IOException {
+                final HttpClientContext clientContext = scope.clientContext;
+                try {
+                    if (entityProducer != null && !entityProducer.isRepeatable()) {
+                        log.debug("{}: cannot retry non-repeatable request", exchangeId);
+                        return asyncExecCallback.handleResponse(response, entityDetails);
+                    }
+                    if (retryStrategy.retryRequest(response, execCount, clientContext)) {
+                        // TODO What to do here?
+                        //response.close();
+                        final TimeValue nextInterval =
+                                retryStrategy.getRetryInterval(response, execCount, clientContext);
+                        if (TimeValue.isPositive(nextInterval)) {
+                            try {
+                                log.debug("{}: wait for {}", exchangeId, nextInterval);
```

Comment:
@michael-o Please use `log.isDebugEnabled()` as the rest of the library in order to avoid unnecessary object allocation.

Reply:
Sorry, what objects? both `exchangeId` and `nextInterval` already exist. SLF4J will only send the log message to the underlying framwork if debug is actually enabled. From my understanding `is..Enabled()` is redundant here. Am I wrong?

Reply:
> Am I wrong?

@michael-o I think you are. There will be an intermediate array object created on the heap. Please use `is..Enabled()` as the rest of the library.

Reply:
While this is right, there is an overload for one and two args: http://www.slf4j.org/apidocs/org/slf4j/Logger.html#debug(java.lang.String,java.lang.Object,java.lang.Object)

So no array object creation here?

Reply:
@ok2c Did you have a chance to check the link? Do you still want me to add the `is..Enabled()` test?

Reply:
@michael-o I still do. The overloaded method just highlights the problem. Will there be an overloaded method for 3 parameters, 4 parameters? What about 5 parameters? `is..Enabled()` check ensures there will be no unnecessary work done, of any kind, ever.

Reply:
No, Ceki says that most log statements have upto 2 args. Therefore the overloads, the rest is with varags. I will add the test to make it consistent with the rest.